### DISCUSSION TITLE
2007/06/28/labview-%E5%AF%B9%E5%A4%9A%E6%A0%B8-cpu-%E7%9A%84%E6%94%AF%E6%8C%81

### DISCUSSION THREAD

**Discussion by giscus** [ruanqizhen/blog#169] [Announcements]
# 2007/06/28/labview-%E5%AF%B9%E5%A4%9A%E6%A0%B8-cpu-%E7%9A%84%E6%94%AF%E6%8C%81

以前，在计算机领域有个摩尔定律，是说每一年半，CPU 的主频都会提高一倍。但是近几年这个定律在CPU主频上已经失效了，我 4 年前用的计算机 CPU 主频是 2G，我前几天换了一台新电脑，CPU 主频还是 2G。

https://blog.qizhen.xyz/2007/06/28/labview-%E5%AF%B9%E5%A4%9A%E6%A0%B8-cpu-%E7%9A%84%E6%94%AF%E6%8C%81

**Comment by ruanqizhen** [comment]
之前博客下的留言：

Qizhen
2009年10月27日 09:18 编辑
Hongbo, 非常感谢你分享的经验。对于读者来说，你的实际经验比我的理论推导更加有价值。再次表示感谢，并希望我们以后能多多交流。

回复
Hongbo
2009年10月27日 01:49 编辑
阮前辈您好！小弟潜水了一段时间，从您的blog里面学到了很多实用的东西。在这里发表一点个人关于多核CPU的拙见。现在（2009年10月）的电脑主流恐怕不只2个或4个核了，如何用多核CPU实时采集和处理大流量的数据是个很有意义的话题，这里我提供一点自己的经验。我的实验系统用一块PXI-5122 digitizer采集图像数据（这个图像比较特别，是由光电倍增管产生的连续信号，需要在软件里重新组合像素位置并生成二维图像）。两个通道各25MHz的采样频率，8比特，这样PXI总线数据流量50MB/S，这个任务只能用一个CPU核作。有个瓶颈是PXI总线传输速率虽然很快（800MB/S）但是启动传输需要一个固定的时间，这样就只能尽可能连续的传输多幅图像*（每幅25毫秒）然后再作像素重建。问题就来自这里，重建像素需要大量的计算时间，单独一个CPU核是没法实时完成这个任务的。还好我有8个CPU，我一开始也尝试按您这里描述的模式手动分工，但是效果并不理想，好像Vista系统自动又重新分配了任务。而且Time loop我的感觉是要比while loop耗费更多的绝对CPU时间，实际测试每个CPU都几乎全负载，还是不能完全实时完成任务。最后我的解决方案是把重建像素的VI的execution属性设置成reentrant和high priority，还有要单独分配内存空间preallocate clone for each instance，编成使用8个同样的VI（也就是8个instance），然后把数据流按图像幅拆成8幅送给这8个VI（同在一个while loop里）并设置好index，输出的图像按index统一放在另开辟好的image序列里（IMAQ create image,循环n次可创建n个image的内存空间）。然后另一个while loop按照index从序列里面依次取出每幅图像并连续显示。这样做的结果是每个CPU都均匀的负载50%左右，除了一个CPU因为要读PXI数据占用了将近90%。我本来想用分7个CPU的方法，另一个单独只作数据采集。但是index看起来就不那么整齐了，所以就让vista自己去分配好了。实际上整个程序除了这部分之外还有很多其他实时处理的内容，不过耗CPU并不多。还有就是把重建好的图像写入AVI格式文件，这个也是通过从image序列里面依次取出图像来完成的，只取决于写硬盘速度。总结一下上面的经验，以下几点可供参考，当然这只是从这一个特例引出来的：1. 对于多核CPU，在Windows系列操作系统里面尽量不要手动指定任务，相信微软做的更好。2. 除了需要严格定时并且有相应的硬件的情况外，尽量用while loop而不是time loop3. 如果一个CPU不能实时完成某个处理大流量数据的VI，把数据平均分拆成n部分，开n个相同的reentrant VI，设置好接受输出数据的内存空间并作好index（非常重要，否则即便VI可以由不同的CPU执行，写内存却要冲突），放在同一个while loop里面。4. 尽量把多个不同的VI放在同一个while loop里面，但是运算周期不一样的VI不要放在同个while loop里以免造成资源浪费。通常情况下，用于界面控制的各种运算处理的VI不需要很高的重复执行率，因此不要和处理数据的VI放一起。多谢砖家点评！

回复
